### ISSUE TITLE
Parallel creation of xml result file

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#2549]
This is a thought issue.

On my users systems, which are very slow, as previously noted (ARM 1Ghz or less), it can take as much as 10x time to build the xml result file as it took to actually run the tests!

Does it make sense to build the xml result file in parallel with the execution of the tests, on a separate thread?  Would this reduce total execution time for other environments as well?

Just interested in thoughts on this for now.

**Comment by mikkelbu** [comment]
No that I've much input at the moment besides it sounds like a good idea, but this issues also sounds like it could be a duplicate of (or included in) #1514.

**Comment by CharliePoole** [comment]
The idea behind the current implementation was that the XML file is only built for runners that need it.

It's entirely possible that this has been lost and that runners not needing the result file are not taking advantage of the possibility of not building it at all. The NUnit GUI for example, gets a full copy of the tests via explore but never asks for the top-level return value, which requires generation of the full XML. It's also possible that changes have gotten in, which cause the framework to always generate the full XML, whether needed or not. Those are two things I'd investigate as potential low-hanging fruit before re-implementing the way the XML is generated.

The `TNode` class is an implementation detail only. It goes back to the need to support multiple platforms without a common way of handling XML. Silverlight was a big factor. Somebody could review what's available in our current (and any likely future) platforms and possibly figure out a different implementation of our `ToXml` and `AddToXml` APIs.

**Comment by CharliePoole** [comment]
IF it turns out that generating XML "as we go" is really desirable (I have doubts) it would have to be done by capturing the `<test-case>` and `<test-suite>` elements that we generate as tests run.  It would have to allow for an open-ended number of currently incomplete higher-level suites at any point in time. It would only need an extra thread if the designer decided to capture the events before the event pump queue, rather than after it. If done after the queue, then the event pump thread could probably do the extra work.

I would not want to see anybody work on this speculatively, however. It requires some performance measurement first, which is what Rob is doing with #2488. It's easy enough to gauge the impact of the XML generation by simply commenting it out temporarily!

**Comment by OmicronPersei** [comment]
What would our options be for measuring performance both on x86 and ARM?

**Comment by CharliePoole** [comment]
External measurement using various runs of nunitlite-runner would be a good start.

An option is to add some internal measurement, controlled by command-line options - or at least the output would be so controlled. The full set of XML is generated lazily through a call to a single property, so timing it would be pretty easy.

**Comment by rprouse** [comment]
@OsirisTerje and I have been working with the Visual Studio Test team to help improve performance of the adapter and NUnit within Visual Studio. We have made several improvements, but the biggest bottleneck remaining is the XML generation. Making the generation Parallel might help or we could analyze the areas that are causing problems and tune those (see #2488).

When we were testing, we found it useful to create very large test suites to make the timings large enough that they are easier to track. @OsirisTerje has a project that generates 10,000 tests or more using T4 templates. It was useful.

I would also like to know that Parallel processing would help. Are you sure that it isn't a memory pressure problem on the smaller devices causing GC cycles or even swapping?

Should we keep this issue open, or is it a duplicate of #2488?

**Comment by CharliePoole** [comment]
@rprouse I don't see this as being at all the same as #2488. According to Navin's comment, there is a lot of time taken in the cross-appdomain calls for the test started and test finished event. That's actually separate from the generation of the XML, which happens before the events can be transmitted. 

I've only been loosely following the discussion, but most of the tests seem to involve other software components, whether the adapter or the console. Granted that our users are affected by the interaction of these components with the framework, I nevertheless think we want to make framework-side-only tests in order to narrow down where the issues are taken.

As mentioned on another issue, I also have a strong suspicion that the adapter, which doesn't need the full XML file, may be inadvertently requesting it anyway by accessing the property that causes it to be generated. The engine design does allow a client that __only__ needs the notifications to get them without ever requesting an aggregated result. Clearly, __not__ doing something that doesn't need to be done is always a good idea when fixing performance issues.

It seems to me that this is easy for us to test because we have an understanding of how the framework works. We also have a strong motivation to make the framework efficient for __all__ use, not just under a particular runner like the adapter.

**Comment by rprouse** [comment]
This issue is marked as an idea but hasn't had contributions in years so I am closing. If anyone is interested in working on this idea, post your interest and the team will consider reopening.
